### ISSUE TITLE
range() overload variant error for type that supports __index__

### ISSUE THREAD

**Issue by ixje** [python/mypy#9755]
**Bug Report**
The documentation for the built-in `range()` function states [ref](https://docs.python.org/3/library/stdtypes.html?highlight=range#range)

>  The arguments to the range constructor must be integers (either built-in int or any object that implements the `__index__` special method).

mypy throws the following error 
> No overload variant of "range" matches argument type "MyInt"

for a custom type with `__index__` implemented

**To Reproduce**
1. Put the following in a file e.g. scratchpad.py
   ```python
   class MyInt:
       def __index__(self):
           return 2
   
   for i in range(MyInt()):
       print(i)
   ```
2. run `mypy scratchpad.py`

**Expected Behavior**
no errors found

**Actual Behavior**
```
scratchpad.py:5: error: No overload variant of "range" matches argument type "MyInt"
scratchpad.py:5: note: Possible overload variant:
scratchpad.py:5: note:     def range(self, stop: int) -> range
scratchpad.py:5: note:     <1 more non-matching overload not shown>
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.790
- Mypy command-line flags: n/a
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.8.3
- Operating system and version: OSX 10.15.7
